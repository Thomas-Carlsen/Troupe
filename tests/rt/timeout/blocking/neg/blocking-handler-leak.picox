let 
    (* set secret to either 0 or 1 *)
    val secret = 1 raisedTo {secret}
    fun wait () = rcv ({}, {secret}, authority, [ hn x => () ])
    val mainPid = self ()
    val h1 = hn 1 => () 
    val h2 = hn 2 => ()

    fun leak i = 
          let val (g1, g2) = 
                if ( (secret > 0) andalso (i = 1))
                   orelse ( (secret <= 0 ) andalso (i = 2))
                   then (h1, h2) else (h2, h1)               
              val _ = receive [g1]              
              val _ = adv i
              val _ = printWithLabels i                     
              val _ = send (mainPid, ())
              val _ = receive [g2]
          in printWithLabels "DONE"
          end       

    (* uncomment to see that the secret is properly labelled *)
    (* val _ = adv secret  *)

    val p1 = spawn ( fn () => leak 1 )
    val p2 = spawn ( fn () => leak 2 )    

    fun batch x = 
     let 
         val _ = printWithLabels x
         val _ = send (p1, x) 
         val _ = send (p2, x)
     in wait ()
     end

    val _ = batch 1
    val _ = batch 2 
in () 
end
