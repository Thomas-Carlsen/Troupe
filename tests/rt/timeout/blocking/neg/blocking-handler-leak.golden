2019-03-07T12:08:28.615Z [RTM] [32minfo[39m: Skipping network creation. Observe that all external IO operations will yield a runtime error.
1@{}%{}
[31mRuntime error in thread 20cfd0d2-5178-48ad-858d-a61de6936896@{}%{}[39m
[31m>> Illegal flow in adv function:[39m
[31m |    pc: {}[39m
[31m | block: {secret}[39m
[31m | value: 1@{}%{}[39m
